### PR TITLE
ISPN-12646 IllegalArgumentException when doing Rolling Upgrades on transactional caches

### PR DESCRIPTION
**Pull request by gustavonalle**
https://issues.redhat.com/browse/ISPN-12646

Forward port from https://github.com/infinispan/infinispan/pull/9007

### CODE REVIEW

**Comment by gustavonalle** [comment]
Failures unrelated

**Comment by tristantarrant** [comment]
Merged from other PR

**Comment by gustavonalle** [comment]
@tristantarrant The other PR didn't have all the commits from this one, I will rebase

**Comment by tristantarrant** [comment]
oops

**Comment by gustavonalle** [comment]
should be ready to merge @tristantarrant

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/ExternalizerIds.java [hunk 1]
```diff
@@ -10,6 +10,6 @@
 
    Integer MIGRATION_TASK = 1900;
    Integer REMOVED_FILTER = 1901;
-   Integer DISCONNECT_REMOTE_STORE = 1902;
-
+   Integer ENTRY_WRITER = 1902;
+   Integer DISCONNECT_REMOTE_STORE = 1903;
```
Comment (pruivo) [inline persistence/remote/src/main/java/org/infinispan/persistence/remote/ExternalizerIds.java:14]: why increasing the DISCONNECT_REMOTE_STORE id? couldn't ENTRY_WRITER be 1093?

Comment (gustavonalle) [reply persistence/remote/src/main/java/org/infinispan/persistence/remote/ExternalizerIds.java:14]: No reason, it was a merge issue, let me put it back

Comment (gustavonalle) [reply persistence/remote/src/main/java/org/infinispan/persistence/remote/ExternalizerIds.java]: fixed

**Review by pruivo — APPROVED** [review]

File: core/src/main/java/org/infinispan/container/entries/VersionedRepeatableReadEntry.java [hunk 2]
```diff
@@ -102,6 +103,9 @@ private boolean skewed(IncrementableEntryVersion prevVersion, EntryVersion versi
          }
          IncrementableEntryVersion prevVersion = versionFromEntry(ice);
          if (prevVersion == null) {
+            if (rollingUpgrade) {
+               return versionGenerator.nonExistingVersion();
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/container/entries/VersionedRepeatableReadEntry.java:107]: Wouldn't it be better to return `nonExistingVersion()` every time, without adding the `rollingUpgrade` parameter?

Comment (gustavonalle) [reply core/src/main/java/org/infinispan/container/entries/VersionedRepeatableReadEntry.java:107]: I'm not sure, wdyt @pruivo ?

Comment (pruivo) [reply core/src/main/java/org/infinispan/container/entries/VersionedRepeatableReadEntry.java:107]: we never have `prevVersion==null` except for the rolling upgrade scenario. 
Hot Rod isn't aware of the internal version and Infinispan stores the entry from RemoteStore as is. 
I would leave the exception there at least to make sure there are no errors.

**Comment by pruivo** [comment]
integrated! thanks @gustavonalle !

**Pull request merged**